*** node_modules/enketo-core/src/js/output.js	2017-04-20 00:28:05.000000000 +0200
--- static/js/enketo/core/output.js	2017-08-23 15:35:19.000000000 +0200
***************
*** 64,69 ****
--- 64,78 ----
              }
              if ( $output.text() !== val ) {
                  $output.text( val );
+ 
+ // MEDIC CHANGES : BEGIN
+                 // For <output>s that are children of choice-groups, trigger
+                 // dependent calculations in case one is using a jr:choice-name.
+                 var owName = $output.closest( '.option-wrapper' ).find( 'input' ).attr( 'name' );
+                 if ( owName ) {
+                     that.form.calc.update( owName );
+                 }
+ // MEDIC CHANGES : END
              }
          } );
      }
