*** node_modules/bootstrap-daterangepicker/daterangepicker.js	2017-01-18 07:47:35.000000000 +1300
--- /home/gareth/Desktop/daterangepicker.js	2017-03-08 22:54:51.689697001 +1300
***************
*** 1295,1300 ****
--- 1295,1303 ----
  
          clickDate: function(e) {
  
+             // MONKEY PATCH BY GRUNT: Needed for the mobile version.
+             this.element.trigger('mm.dateSelected.daterangepicker', this);
+ 
              if (!$(e.target).hasClass('available')) return;
  
              var title = $(e.target).attr('data-title');
