*** copy.orig.js	2022-05-12 16:03:34.887962008 +0200
--- copy.js	2022-05-12 16:04:06.784553121 +0200
***************
*** 76,81 ****
--- 76,89 ----
  
        filePair.src.forEach(function(src) {
          src = unixifyPath(src);
+ 
+         if (filePair.orig.follow) {
+           var isSymlink = fs.lstatSync(src).isSymbolicLink();
+           if (isSymlink) {
+             src = fs.readlinkSync(src);
+           }
+         }
+ 
          var dest = unixifyPath(filePair.dest);
  
          if (detectDestType(dest) === 'directory') {
