*** node_modules/moment/locale/hi.js	1985-10-26 20:15:00.000000000 +1200
--- /home/gareth/Desktop/hi.js	2018-04-09 11:38:14.716147244 +1200
***************
*** 7,38 ****
     factory(global.moment)
  }(this, (function (moment) { 'use strict';
  
- 
-     var symbolMap = {
-         '1': '१',
-         '2': '२',
-         '3': '३',
-         '4': '४',
-         '5': '५',
-         '6': '६',
-         '7': '७',
-         '8': '८',
-         '9': '९',
-         '0': '०'
-     },
-     numberMap = {
-         '१': '1',
-         '२': '2',
-         '३': '3',
-         '४': '4',
-         '५': '5',
-         '६': '6',
-         '७': '7',
-         '८': '8',
-         '९': '9',
-         '०': '0'
-     };
- 
      var hi = moment.defineLocale('hi', {
          months : 'जनवरी_फ़रवरी_मार्च_अप्रैल_मई_जून_जुलाई_अगस्त_सितम्बर_अक्टूबर_नवम्बर_दिसम्बर'.split('_'),
          monthsShort : 'जन._फ़र._मार्च_अप्रै._मई_जून_जुल._अग._सित._अक्टू._नव._दिस.'.split('_'),
--- 7,12 ----
***************
*** 73,86 ****
              yy : '%d वर्ष'
          },
          preparse: function (string) {
!             return string.replace(/[१२३४५६७८९०]/g, function (match) {
!                 return numberMap[match];
!             });
          },
          postformat: function (string) {
!             return string.replace(/\d/g, function (match) {
!                 return symbolMap[match];
!             });
          },
          // Hindi notation for meridiems are quite fuzzy in practice. While there exists
          // a rigid notion of a 'Pahar' it is not used as rigidly in modern Hindi.
--- 47,56 ----
              yy : '%d वर्ष'
          },
          preparse: function (string) {
!             return string;
          },
          postformat: function (string) {
!             return string;
          },
          // Hindi notation for meridiems are quite fuzzy in practice. While there exists
          // a rigid notion of a 'Pahar' it is not used as rigidly in modern Hindi.
