*** node_modules/pouchdb/lib/index-browser.js	2016-05-10 15:57:49.654362714 +1200
--- index-browser-updated.js	2016-05-10 15:58:11.077862181 +1200
***************
*** 3892,3906 ****
      if (changes.results.length > 0) {
        changesOpts.since = changes.last_seq;
        getChanges();
      } else {
!       if (continuous) {
!         changesOpts.live = true;
!         getChanges();
        } else {
!         changesCompleted = true;
        }
      }
-     processPendingBatch(true);
    }
  
  
--- 3892,3924 ----
      if (changes.results.length > 0) {
        changesOpts.since = changes.last_seq;
        getChanges();
+       processPendingBatch(true);
      } else {
! 
!       var complete = function () {
!         if (continuous) {
!           changesOpts.live = true;
!           getChanges();
!         } else {
!           changesCompleted = true;
!         }
!         processPendingBatch(true);
!       };
! 
!       // update the checkpoint so we start from the right seq next time
!       if (!currentBatch && changes.last_seq > last_seq) {
!         writingCheckpoint = true;
!         checkpointer.writeCheckpoint(changes.last_seq,
!             session).then(function () {
!           writingCheckpoint = false;
!           result.last_seq = last_seq = changes.last_seq;
!           complete();
!         })
!         .catch(onCheckpointError);
        } else {
!         complete();
        }
      }
    }
  
  
