*** node_modules/pouchdb-original/lib/index.js  2016-03-18 09:58:47.000000000 +1300
--- node_modules/pouchdb/lib/index.js 2016-04-18 09:36:28.339098861 +1200
***************
*** 3167,3172 ****
--- 3173,3181 ----
          bulkGetInfo.docs.forEach(function (doc) {
            if (doc.ok) {
              resultDocs.push(doc.ok);
+           // MONKEY PATCH BY GRUNT: retry failed batches
+           } else if(typeof doc.error !== 'undefined') {
+             throw new Error('Bad doc in batch: ' + JSON.stringify(doc));
            }
          });
        });
