*** ./pouchdb-adapter-idb/src/index.js	2017-03-08 16:25:19.192137941 +1300
--- index-patched.js	2017-03-08 16:25:44.819510389 +1300
***************
*** 1924,1929 ****
--- 1924,1930 ----
    var isSafari = typeof openDatabase !== 'undefined' &&
      /(Safari|iPhone|iPad|iPod)/.test(navigator.userAgent) &&
      !/Chrome/.test(navigator.userAgent) &&
+     !/TECNO/.test(navigator.userAgent) &&
      !/BlackBerry/.test(navigator.platform);
  
    // some outdated implementations of IDB that appear on Samsung
