*** unpatched	2021-10-12 11:53:06.753876350 -0500
--- patched  	2021-10-12 11:55:37.011514348 -0500
***************
*** 58,64 ****
                       * and the dependent node is inside the same repeat, we can prevent the expensive index determination
                       */
                      const dataNodeName = ( name.lastIndexOf( '/' ) !== -1 ) ? name.substring( name.lastIndexOf( '/' ) + 1 ) : name;
!                     const dataNode = this.form.model.node( updated.repeatPath, updated.repeatIndex ).getElement().querySelector( dataNodeName );
                      props.index = dataNodes.indexOf( dataNode );
                      this._updateCalc( control, props, emptyNonRelevant );
                  } else if ( control.type === 'hidden' ) {
--- 58,65 ----
                       * and the dependent node is inside the same repeat, we can prevent the expensive index determination
                       */
                      const dataNodeName = ( name.lastIndexOf( '/' ) !== -1 ) ? name.substring( name.lastIndexOf( '/' ) + 1 ) : name;
!                     const childNodeList = this.form.model.node( updated.repeatPath, updated.repeatIndex ).getElement().querySelectorAll( dataNodeName );
!                     const dataNode = Array.from( childNodeList ).filter( node => dataNodes.includes( node ) )[0];
                      props.index = dataNodes.indexOf( dataNode );
                      this._updateCalc( control, props, emptyNonRelevant );
                  } else if ( control.type === 'hidden' ) {
