*** webapp/node_modules/enketo-core/src/js/page.js	2019-09-10 13:09:07.407455858 +1200
--- webapp/node_modules/enketo-core/src/js/page.new.js	2019-09-10 13:09:35.278897963 +1200
***************
*** 274,280 ****
              .eq( 0 )
              .trigger( 'fakefocus' );
  
!         pageEl.scrollIntoView();
      },
      toggleButtons: function( index ) {
          var i = index || this.getCurrentIndex(),
--- 274,283 ----
              .eq( 0 )
              .trigger( 'fakefocus' );
  
!         if (pageEl) {
!             pageEl.scrollIntoView();
!             console.warn('No active pages found. Make sure your form has at least one relevant input.');
!         }
      },
      toggleButtons: function( index ) {
          var i = index || this.getCurrentIndex(),
