*** unpatched	1985-10-26 03:15:00.000000000 -0500
--- patched	2022-04-11 15:11:01.396731595 -0500
***************
*** 800,805 ****
--- 800,808 ----
              const path = that.input.getName(node);
              let target;
  
+             // /inputs is ALWAYS relevant #4875
+             if(/\/inputs$/.test(path)) return;
+ 
              /*
               * Copied from relevant.js:
               *
