*** unpatched	2022-04-05 11:46:20.970352329 -0500
--- patched	2022-04-05 11:50:47.743949897 -0500
***************
*** 216,222 ****
  
          if ( neverEnabled || this.selfRelevant( branchNode ) || forceClearNonRelevant ) {
              changed = true;
!             if ( forceClearNonRelevant ) {
                  this.clear( branchNode, path );
              }
  
--- 216,223 ----
  
          if ( neverEnabled || this.selfRelevant( branchNode ) || forceClearNonRelevant ) {
              changed = true;
!             // /inputs is ALWAYS relevant #4875
!             if ( forceClearNonRelevant && !path.endsWith('/inputs') ) {
                  this.clear( branchNode, path );
              }
  
