diff --git a/node_modules/enketo-core/src/js/form.js b/node_modules/enketo-core/src/js/form.js
index 62003af..c9f749c 100644
--- a/node_modules/enketo-core/src/js/form.js
+++ b/node_modules/enketo-core/src/js/form.js
@@ -939,6 +939,10 @@ Form.prototype.getDataStrWithoutIrrelevantNodes = function () {
             const path = that.input.getName(node);
             let target;

+            // CHT-CORE PATCH
+            // /inputs is ALWAYS relevant #4875
+            if(/\/inputs$/.test(path)) return;
+
             /*
              * Copied from relevant.js:
              *
diff --git a/node_modules/enketo-core/src/js/relevant.js b/node_modules/enketo-core/src/js/relevant.js
index 463af2c..d55a32b 100644
--- a/node_modules/enketo-core/src/js/relevant.js
+++ b/node_modules/enketo-core/src/js/relevant.js
@@ -546,7 +546,9 @@ export default {
         ) {
             changed = true;

-            if (forceClearNonRelevant) {
+            // CHT PATCH
+            // /inputs is ALWAYS relevant #4875
+            if (forceClearNonRelevant  && !path.endsWith('/inputs')) {
                 this.clear(branchNode, path);
             }

