*** unpatched	1985-10-26 03:15:00.000000000 -0500
--- patched	2022-04-11 15:14:55.118064979 -0500
***************
*** 487,494 ****
              forceClearNonRelevant
          ) {
              changed = true;
! 
!             if (forceClearNonRelevant) {
                  this.clear(branchNode, path);
              }
  
--- 487,494 ----
              forceClearNonRelevant
          ) {
              changed = true;
!             // /inputs is ALWAYS relevant #4875
!             if (forceClearNonRelevant && !/\/inputs$/.test(path)) {
                  this.clear(branchNode, path);
              }
  
