diff --git a/node_modules/messageformat/lib/plurals.js b/node_modules/messageformat/lib/plurals.js
index f3433a2..ec70b91 100644
--- a/node_modules/messageformat/lib/plurals.js
+++ b/node_modules/messageformat/lib/plurals.js
@@ -38,15 +38,21 @@ function wrapPluralFunc(lc, pf, pluralKeyChecks) {
   return fn;
 }
 
+function defaultPluralFunc(n, ord) {
+  if (ord) return 'other';
+  return 'other';
+}
+
 function getPlural(locale, {
   pluralKeyChecks
 }) {
-  for (let lc = String(locale); lc; lc = lc.replace(/[-_]?[^-_]*$/, '')) {
+  let lc;
+  for (lc = String(locale); lc; lc = lc.replace(/[-_]?[^-_]*$/, '')) {
     const pf = _plurals.default[lc];
     if (pf) return wrapPluralFunc(lc, pf, pluralKeyChecks);
   }
 
-  throw new Error('Localisation function not found for locale ' + JSON.stringify(locale));
+  return wrapPluralFunc(lc, defaultPluralFunc, pluralKeyChecks);
 }
 
 function getAllPlurals({
