*** /home/gareth/Desktop/plurals.js	2019-06-25 16:28:59.920684743 +1200
--- /home/gareth/Desktop/plurals.patched.js	2019-06-25 16:44:23.967022810 +1200
***************
*** 38,50 ****
    return fn;
  }
  
  function getPlural(locale, noPluralKeyChecks) {
!   for (let lc = String(locale); lc; lc = lc.replace(/[-_]?[^-_]*$/, '')) {
      const pf = _plurals.default[lc];
      if (pf) return wrapPluralFunc(lc, pf, noPluralKeyChecks);
    }
  
!   throw new Error('Localisation function not found for locale ' + JSON.stringify(locale));
  }
  
  function getAllPlurals(noPluralKeyChecks) {
--- 38,56 ----
    return fn;
  }
  
+ function defaultPluralFunc(n, ord) {
+   if (ord) return 'other';
+   return 'other';
+ }
+ 
  function getPlural(locale, noPluralKeyChecks) {
!   let lc;
!   for (lc = String(locale); lc; lc = lc.replace(/[-_]?[^-_]*$/, '')) {
      const pf = _plurals.default[lc];
      if (pf) return wrapPluralFunc(lc, pf, noPluralKeyChecks);
    }
  
!   return wrapPluralFunc(lc, defaultPluralFunc, noPluralKeyChecks);
  }
  
  function getAllPlurals(noPluralKeyChecks) {
