*** nool.parser.js	Mon Jul  6 12:01:37 2020
--- nool.parser-copy.js	Mon Jul  6 11:52:27 2020
***************
*** 7,13 ****
  
  var parse = function (src, keywords, context) {
      var orig = src;
!     src = src.replace(/\/\/(.*)/g, "").replace(/\r\n|\r|\n/g, " ");
  
      var blockTypes = new RegExp("^(" + keys(keywords).join("|") + ")"), index;
      while (src && (index = utils.findNextTokenIndex(src)) !== -1) {
--- 7,15 ----
  
  var parse = function (src, keywords, context) {
      var orig = src;
!     src = src.replace(/\/\*[\s\S]*?\*\/|([^:]|^)\/\/.*$/gm, " "+/\/\*[\s\S]*?\*\/|([^:]|^)\/\/.*$/gm)
!     .replace(/\/\*[\s\S]*?\*\/|([^:]|^)\/\/.*$/gm, " ")
!     .replace(/\r\n|\r|\n/g, " ");
  
      var blockTypes = new RegExp("^(" + keys(keywords).join("|") + ")"), index;
      while (src && (index = utils.findNextTokenIndex(src)) !== -1) {
