*** nool.parser.js	Wed Jul  8 13:05:00 2020
--- nool.parser-copy.js	Wed Jul  8 13:06:01 2020
***************
*** 7,13 ****
  
  var parse = function (src, keywords, context) {
      var orig = src;
!     src = src.replace(/\/\/(.*)/g, "").replace(/\r\n|\r|\n/g, " ");
  
      var blockTypes = new RegExp("^(" + keys(keywords).join("|") + ")"), index;
      while (src && (index = utils.findNextTokenIndex(src)) !== -1) {
--- 7,13 ----
  
  var parse = function (src, keywords, context) {
      var orig = src;
!     src = src.replace(/\/\*[\s\S]*?\*\/|([^:]|^)\/\/.*$/gm, " ").replace(/\r\n|\r|\n/g, " ");
  
      var blockTypes = new RegExp("^(" + keys(keywords).join("|") + ")"), index;
      while (src && (index = utils.findNextTokenIndex(src)) !== -1) {
