*** webapp/node_modules/pouchdb-browser/lib/index.js	1985-10-26 20:15:00.000000000 +1200
--- webapp/node_modules/pouchdb-browser/lib/index-new.js	2021-05-17 15:58:05.654751125 +1200
***************
*** 876,891 ****
  // the diffFun tells us what delta to apply to the doc.  it either returns
  // the doc, or false if it doesn't need to do an update after all
  function upsert(db, docId, diffFun) {
!   return new Promise(function (fulfill, reject) {
!     db.get(docId, function (err, doc) {
!       if (err) {
!         /* istanbul ignore next */
!         if (err.status !== 404) {
!           return reject(err);
!         }
!         doc = {};
        }
! 
        // the user might change the _rev, so save it for posterity
        var docRev = doc._rev;
        var newDoc = diffFun(doc);
--- 876,889 ----
  // the diffFun tells us what delta to apply to the doc.  it either returns
  // the doc, or false if it doesn't need to do an update after all
  function upsert(db, docId, diffFun) {
!   return db.get(docId)
!     .catch(function (err) {
!       if (err.status !== 404) {
!         throw err;
        }
!       return {};
!     })
!     .then(function (doc) {
        // the user might change the _rev, so save it for posterity
        var docRev = doc._rev;
        var newDoc = diffFun(doc);
***************
*** 893,908 ****
        if (!newDoc) {
          // if the diffFun returns falsy, we short-circuit as
          // an optimization
!         return fulfill({updated: false, rev: docRev});
        }
  
        // users aren't allowed to modify these values,
        // so reset them here
        newDoc._id = docId;
        newDoc._rev = docRev;
!       fulfill(tryAndPut(db, newDoc, diffFun));
      });
-   });
  }
  
  function tryAndPut(db, doc, diffFun) {
--- 891,905 ----
        if (!newDoc) {
          // if the diffFun returns falsy, we short-circuit as
          // an optimization
!         return {updated: false, rev: docRev};
        }
  
        // users aren't allowed to modify these values,
        // so reset them here
        newDoc._id = docId;
        newDoc._rev = docRev;
!       return tryAndPut(db, newDoc, diffFun);
      });
  }
  
  function tryAndPut(db, doc, diffFun) {
